src/plugins/dashboard/public/application/utils/create_dashboard_container.tsx

```diff
@@ -0,0 +1,541 @@
+import { cloneDeep, isEqual, uniqBy } from 'lodash';
+import { i18n } from '@osd/i18n';
+import { EMPTY, Observable, Subscription, merge, pipe } from 'rxjs';
+import {
+  catchError,
+  distinctUntilChanged,
+  filter,
+  map,
+  mapTo,
+  startWith,
+  switchMap,
+} from 'rxjs/operators';
+import deepEqual from 'fast-deep-equal';
+
+import { IndexPattern, opensearchFilters } from '../../../../data/public';
+import {
+  DASHBOARD_CONTAINER_TYPE,
+  DashboardContainer,
+  DashboardContainerInput,
+  DashboardPanelState,
+} from '../embeddable';
+import {
+  ContainerOutput,
+  EmbeddableFactoryNotFoundError,
+  EmbeddableInput,
+  ViewMode,
+  isErrorEmbeddable,
+  openAddPanelFlyout,
+} from '../../embeddable_plugin';
+import {
+  convertPanelStateToSavedDashboardPanel,
+  convertSavedDashboardPanelToPanelState,
+} from '../lib/embeddable_saved_object_converters';
+import { DashboardEmptyScreen, DashboardEmptyScreenProps } from '../dashboard_empty_screen';
+import {
+  DashboardAppState,
+  DashboardAppStateContainer,
+  DashboardServices,
+  SavedDashboardPanel,
+} from '../../types';
+import { getSavedObjectFinder } from '../../../../saved_objects/public';
+import { DashboardConstants } from '../../dashboard_constants';
+import { SavedObjectDashboard } from '../../saved_dashboards';
+import { migrateLegacyQuery } from '../lib/migrate_legacy_query';
+import { Dashboard } from '../../dashboard';
+
+export const createDashboardContainer = async ({
+  services,
+  savedDashboard,
+  appState,
+}: {
+  services: DashboardServices;
+  savedDashboard?: SavedObjectDashboard;
+  appState?: DashboardAppStateContainer;
+}) => {
+  const { embeddable } = services;
+
+  const dashboardFactory = embeddable.getEmbeddableFactory<
+    DashboardContainerInput,
+    ContainerOutput,
+    DashboardContainer
+  >(DASHBOARD_CONTAINER_TYPE);
+
+  if (!dashboardFactory) {
+    throw new EmbeddableFactoryNotFoundError('dashboard');
+  }
+
+  try {
+    if (appState) {
+      const appStateData = appState.getState();
+      const initialInput = getDashboardInputFromAppState(
+        appStateData,
+        services,
+        savedDashboard?.id
+      );
+
+      const incomingEmbeddable = services.embeddable
+        .getStateTransfer(services.scopedHistory)
+        .getIncomingEmbeddablePackage();
+
+      if (
+        incomingEmbeddable?.embeddableId &&
+        initialInput.panels[incomingEmbeddable.embeddableId]
+      ) {
+        const initialPanelState = initialInput.panels[incomingEmbeddable.embeddableId];
+        initialInput.panels = {
+          ...initialInput.panels,
+          [incomingEmbeddable.embeddableId]: {
+            gridData: initialPanelState.gridData,
+            type: incomingEmbeddable.type,
+            explicitInput: {
+              ...initialPanelState.explicitInput,
+              ...incomingEmbeddable.input,
+              id: incomingEmbeddable.embeddableId,
+            },
+          },
+        };
+      }
```

Comment:
Im assuming that this is when the user is redirected to dashboards from another app, e.g. save and return from visualize right?